api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ImmunizationTranslatorImpl.java

```diff
@@ -0,0 +1,215 @@
+		openMrsImmunization.getGroupMembers().stream().forEach(obs -> {
+			obs.setPerson(patient);
+			obs.setLocation(location);
+			obs.setEncounter(encounter.orElse(newEncounter));
+		});
+		
+		Map<String, Obs> members = getObsMembersMap(openMrsImmunization);
+		
+		Coding coding = fhirImmunization.getVaccineCode().getCoding().stream()
+		        .filter(code -> StringUtils.isEmpty(code.getSystem())).reduce((code1, code2) -> {
+			        throw new IllegalArgumentException("Multiple system-less coding found for the immunization's vaccine: "
+			                + code1.getCode() + " and " + code2.getCode()
+			                + ". No unique system concept could be identified as the coded answer.");
+		        }).get();
+		members.get(ciel984).setValueCoded(getConceptService().getConceptByUuid(coding.getCode()));
+		
+		members.get(ciel1410).setValueDatetime(fhirImmunization.getOccurrenceDateTimeType().getValue());
+		
+		if (CollectionUtils.size(fhirImmunization.getProtocolApplied()) != 1) {
+			throw new IllegalArgumentException(
+			        "Either no protocol applied was found or multiple protocols applied were found. Only strictly one protocol is currently supported for each immunization.");
+		}
+		ImmunizationProtocolAppliedComponent protocolApplied = fhirImmunization.getProtocolApplied().get(0);
+		members.get(ciel1418).setValueNumeric(protocolApplied.getDoseNumberPositiveIntType().getValue().doubleValue());
+		
+		members.get(ciel1419).setValueText(fhirImmunization.getManufacturer().getDisplay());
+		
+		members.get(ciel1420).setValueText(fhirImmunization.getLotNumber());
+		
+		members.get(ciel165907).setValueDatetime(fhirImmunization.getExpirationDate());
+		
+		return openMrsImmunization;
+	}
+	
+	@Override
+	public Immunization toFhirResource(Obs openMrsImmunization) {
+		validateImmunizationObsGroup(openMrsImmunization);
+		
+		Immunization immunization = new Immunization();
+		immunization.setId(openMrsImmunization.getUuid());
+		immunization.setStatus(ImmunizationStatus.COMPLETED);
+		immunization.setPatient(
+		    new Reference().setType(PATIENT).setReference(PATIENT + "/" + openMrsImmunization.getPerson().getUuid()));
+		immunization.setEncounter(new Reference().setType(ENCOUNTER)
+		        .setReference(ENCOUNTER + "/" + openMrsImmunization.getEncounter().getVisit().getUuid()));
+		immunization.setPerformer(Arrays.asList(new ImmunizationPerformerComponent(new Reference().setType(PRACTITIONER)
+		        .setReference(PRACTITIONER + "/" + getAdministeringProvider(openMrsImmunization).getUuid()))));
+		
+		Map<String, Obs> members = getObsMembersMap(openMrsImmunization);
+		
+		CodeableConcept codeableConcept = new CodeableConcept();
+		Coding coding = new Coding();
+		coding.setCode(members.get(ciel984).getValueCoded().getUuid());
+		coding.setDisplay(members.get(ciel984).getValueCoded().getFullySpecifiedName(Context.getLocale()).getName());
+		codeableConcept.addCoding(coding);
+		immunization.setVaccineCode(codeableConcept);
```

Comment:
For at least a couple of these, you could use the `ObservationValueTranslator`, though it won't work for the Immunization.protocolApplied

Reply:
I actually could only do it for the vaccine _occurrence_ [here](https://github.com/openmrs/openmrs-module-fhir2/pull/264/files#diff-56eb3743524091606fd9093205f0bc07R197) because it takes a `Type` as argument. For all those that do not take a `Type` I'm not sure if there is a recommended way to do it?